### DISCUSSION TITLE
How to serve different schemas based on incoming request?

### DISCUSSION THREAD

**Discussion by dotansimha** [dotansimha/envelop#144] [General]
I think there could be several approaches here, but I'm still not sure which one is the ideal / recommended. 

* Create multiple `envelop`s with share plugins, and just different `useSchema` calls, then allow users to do the schema routing.
* Create `useSchemaRouting` plugin to allow developers to write a custom function that manages the routing. This will probably will be just a call to `setSchema`? Or probably allow developers to be part of the `get schema()`?
* Something else?

**Comment by dotansimha** [comment]
@n1ru4l thoughts?

**Reply by n1ru4l** [reply]
One more thing is that people might not only want to use a different schema based on the request. I think it should probably be possible to customize any of the following params which are passed to execute/subscribe based on the request 🤔

```ts
export interface ExecutionArgs {
  schema: GraphQLSchema;
  document: DocumentNode;
  rootValue?: unknown;
  contextValue?: unknown;
  variableValues?: Maybe<{ [key: string]: unknown }>;
  operationName?: Maybe<string>;
  fieldResolver?: Maybe<GraphQLFieldResolver<any, any>>;
  typeResolver?: Maybe<GraphQLTypeResolver<any, any>>;
}
```

https://github.com/graphql/graphql-js/blob/917befd7a54fe99d6141fa0ed92bc2b82544625f/src/execution/execute.d.ts#L91-L100

In `graphql-ws` this is possible via the onSubscribe option: https://github.com/enisdenjo/graphql-ws/blob/9103123e32975e615ff5381052753f5e96a2e471/src/server.ts#L278-L285

Maybe it would make the most sense to allow injecting some custom payload into all the phases, so the plugins could use that logic `getEnvelope(context)`.

**Comment by dotansimha** [comment]
This can be implemented now with `useLazyLoadedSchema` plugin :)
